common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(NonZeroUsize, NonZeroUsize) -> Arc<Mutex<Bucket>>>(
+        &self,
+        key: Key,
+        bucket_create: F,
+    ) -> Arc<Mutex<Bucket>> {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        let maybe_value = { self.buckets.read().get(&key).cloned() };
+
+        if let Some(value) = maybe_value {
+            value
+        } else {
+            let size = self.default_bucket_size;
+            let rate = self.default_fill_rate;
+            // Write in a bucket, but make sure again that it isn't there first
+            self.buckets
+                .write()
+                .entry(key)
+                .or_insert_with(|| bucket_create(size, rate))
+                .clone()
+        }
+    }
+
+    /// Garbage collects a single key, if we know what it is
+    pub fn garbage_collect_key(&self, key: &Key) -> bool {
+        let mut write = self.buckets.write();
+        let arc = write.get(key);
+        let remove = arc.map(|arc| Arc::strong_count(arc) <= 1).unwrap_or(false);
+
+        if remove {
+            write.remove(key);
+        }
+
+        remove
+    }
+}
+
+/// A token bucket object that keeps track of everything related to a key
+/// This can be used as a standalone rate limiter; however, to make it more useful
+/// it should be wrapped in an `Arc` and a `Mutex` to be shared across threads.
+#[derive(Debug)]
+pub struct Bucket {
+    /// The current number of available tokens to be used
+    tokens: usize,
+    /// Maximum number of `tokens` in the bucket
+    size: NonZeroUsize,
+    /// The fill rate of the bucket (`tokens/s`).  Amount added to `tokens` on a `refill`
+    rate: NonZeroUsize,
+    /// The last time buckets were refilled, to keep track of for amount to refill
+    last_refresh_time: Instant,
+    /// Determines whether the rate limiting should be ignored, useful for testing
+    open: bool,
+}
+
+impl Bucket {
+    pub fn new(size: NonZeroUsize, rate: NonZeroUsize) -> Self {
+        assert!(
+            size >= rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            tokens: size.get(),
+            size,
+            rate,
+            last_refresh_time: Instant::now(),
+            open: false,
+        }
+    }
+
+    /// A fully open rate limiter, to allow for ignoring rate limiting for tests
+    pub fn open() -> Self {
+        let max = NonZeroUsize::new(std::usize::MAX).unwrap();
+        Self {
+            tokens: std::usize::MAX,
+            size: max,
+            rate: max,
+            last_refresh_time: Instant::now(),
+            open: true,
+        }
+    }
+
+    /// Refill tokens based on how many seconds have passed since last refresh
+    pub(crate) fn refill(&mut self) {
+        let num_intervals = self.last_refresh_time.elapsed().as_secs();
```

Comment:
how would we simulate / test this?

Reply:
It's currently tested in the unit tests below.  With a mock time service, it then won't have to wait seconds.